contracts/StakingBurner.vy

```diff
@@ -0,0 +1,111 @@
+# @version 0.2.4
+"""
+@title Staking Burner
+@author Curve Finance
+@license MIT
+@notice Used for share profit
+"""
+
+
+from vyper.interfaces import ERC20
+
+
+interface Pool:
+    def coins(i: uint256) -> address: view
+    def exchange(i: int128, j: int128, dx: uint256, min_dy: uint256): nonpayable
+
+
+interface Gauge:
+    def integrate_fraction(_address: address) -> uint256: view
+    def integrate_inv_supply_of(_address: address) -> uint256: view
+
+
+SHARE: constant(decimal) = 0.5
+
+
+pool: public(address)
+pool_proxy: public(address)
+ve_holder: public(address)
+burned: public(uint256)
+burned_of: public(HashMap[address, uint256])
+gauge: public(address)
+integrate_fraction: public(HashMap[address, uint256])
+integrate_inv_supply_of: public(HashMap[address, uint256])
+
+
+@external
+def __init__(_ve_holder: address, _pool: address, _pool_proxy: address, _gauge: address):
+    self.ve_holder = _ve_holder
+    self.pool = _pool
+    self.pool_proxy = _pool_proxy
+    self.gauge = _gauge
+
+
+@external
+@nonreentrant('lock')
+def burn_coin(_coin: address) -> bool:
+    last_fraction: uint256 = Gauge(self.gauge).integrate_fraction(msg.sender)
+    holder_fraction: uint256 = self.integrate_fraction[msg.sender]
+    assert holder_fraction < last_fraction, "Already burned"
+
+    potential_profit: uint256 = ERC20(_coin).balanceOf(self.pool_proxy) + self.burned
+    ve_profit: uint256 = convert(convert(potential_profit, decimal) * SHARE, uint256)
```

Comment:
I am a little bit uncomfortable with decimals. It will overflow at about $10B. Maybe large enough but who knows (-: This is because decimals are only 128-bit numbers, and they have 10 decimal digits

I usually like to do something like
```
SHARE = 5 * 10**17
...
ve_profit = potential_profit * SHARE / 10**18
```

@iamdefinitelyahuman thoughts?